clicommand/pipeline_upload.go

```diff
@@ -613,24 +613,73 @@ func (cfg *PipelineUploadConfig) parseAndInterpolate(ctx context.Context, src st
 
 // gatherChangedFiles determines changed files in this build.
 func gatherChangedFiles(diffBase string) (mergeBase string, changedPaths []string, err error) {
-	// The --merge-base flag was only added to git-diff recently.
-	mergeBaseOut, err := exec.Command("git", "merge-base", diffBase, "HEAD").Output()
+	// Corporate needs you to find the differences between diffBase and HEAD.
+	diffBaseCommit, err := exec.Command("git", "rev-parse", diffBase).Output()
 	if err != nil {
-		return "", nil, gitMergeBaseError{diffBase: diffBase, wrapped: err}
+		return "", nil, gitRevParseError{arg: diffBase, wrapped: err}
 	}
-	mergeBase = strings.TrimSpace(string(mergeBaseOut))
-
-	gitDiff, err := exec.Command("git", "diff", "--name-only", mergeBase).Output()
+	headCommit, err := exec.Command("git", "rev-parse", "HEAD").Output()
 	if err != nil {
-		return mergeBase, nil, gitDiffError{mergeBase: mergeBase, wrapped: err}
+		return "", nil, gitRevParseError{arg: "HEAD", wrapped: err}
+	}
+	if strings.TrimSpace(string(diffBaseCommit)) == strings.TrimSpace(string(headCommit)) {
+		// They're the same commit, so `git diff` will (correctly) report no
+		// changes between them.
+		// Instead, we can use a git log invocation to get changed files from
+		// the commit, including merge commits. Flag Explainer:
+		// `--first-parent`: when reaching a merge commit, follow the first parent
+		// (the branch that was merged *into*). Implies --diff-merges=first-parent,
+		// which is important here.
+		// `-1`: show only the most recent entry from the log
+		// `--name-only`: show only the name of each changed file in the post-image tree
+		// `--pretty='format:'`:
+		gitLog, err := exec.Command("git", "log", "--first-parent", "-1", "--name-only", "--pretty='format:'").Output()
+		if err != nil {
+			return "", nil, gitLogError{wrapped: err}
+		}
+		changedPaths = strings.Split(string(gitLog), "\n")
```

Comment:
Do we want to assign `mergeBase`? or it means nothing because the `HEAD` does not intend to be merged.

Reply:
Good point. Rather than pretend there is a merge base, I've changed `gatherChangedFiles` to be responsible for logging what it is doing.